pkg/controller/common/license/trial.go

```diff
@@ -9,74 +9,148 @@ import (
 	"crypto/rsa"
 	"crypto/x509"
 	"errors"
+	"fmt"
+	"strconv"
 	"time"
 
 	"github.com/elastic/cloud-on-k8s/pkg/utils/chrono"
-	"github.com/elastic/cloud-on-k8s/pkg/utils/k8s"
 	pkgerrors "github.com/pkg/errors"
 	corev1 "k8s.io/api/core/v1"
 	metav1 "k8s.io/apimachinery/pkg/apis/meta/v1"
+	"k8s.io/apimachinery/pkg/types"
 	"k8s.io/apimachinery/pkg/util/uuid"
 )
 
 const (
 	TrialStatusSecretKey = "trial-status"
 	TrialPubkeyKey       = "pubkey"
+	TrialActivationKey   = "in-trial-activation"
 
 	TrialLicenseSecretName      = "trial.k8s.elastic.co/secret-name"      // nolint
 	TrialLicenseSecretNamespace = "trial.k8s.elastic.co/secret-namespace" // nolint
 )
 
-func InitTrial(c k8s.Client, operatorNamespace string, secret corev1.Secret, l *EnterpriseLicense) (*rsa.PublicKey, error) {
-	if l == nil {
-		return nil, errors.New("license is nil")
+// TrialState captures the in-memory representation of the trial status.
+type TrialState struct {
+	privateKey *rsa.PrivateKey
+	publicKey  *rsa.PublicKey
+}
+
+// NewTrialState creates a new trial state based on a new RSA key pair.
+func NewTrialState() (TrialState, error) {
+	key, err := newTrialKey()
+	if err != nil {
+		return TrialState{}, err
 	}
+	return TrialState{
+		privateKey: key,
+		publicKey:  &key.PublicKey,
+	}, nil
+}
 
-	if err := populateTrialLicense(l); err != nil {
-		return nil, pkgerrors.Wrap(err, "failed to populate trial license")
+// NewTrialStateFromStatus reconstructs trial state from a trial status secret.
+func NewTrialStateFromStatus(trialStatus corev1.Secret) (TrialState, error) {
+	// create new keys if the operator failed just before the trial was started
+	trialActivation, err := strconv.ParseBool(string(trialStatus.Data[TrialActivationKey]))
+	if err == nil && trialActivation {
+		return NewTrialState()
```

Comment:
should we handle a case where the operator starts, sees activation is in progress, but some license secrets already exist out there?

Reply:
Thinking about this some more, I think we got that covered. I see two cases where this might happen:
1. Crash/Restart after generating the trial license:
   * we start trial activation, create status secret
   * we then initialise the license, update license
   * crash/restart of operator: status+license match up and we continue to complete trial activation
2. License exists from a previous installation or similar
   * we start trial activation, create status secret
   * existing license + status mismatch we annotate the license
   * operator crashes at any point after that
   * existing license+status still mismatch 
   * we could now delete the license/re-initialise it but my thinking is we should leave that to the user. What do you think?

Reply:
> Crash/Restart after generating the trial license:
we start trial activation, create status secret
we then initialise the license, update license
crash/restart of operator: status+license match up and we continue to complete trial activation

when that happens, aren't we regenerating a new pubkey, in which case status+license don't match up?

Reply:
Would it be better to check existing licenses in that situation?
* operator starts and sees the status reports activation in progress
    * check existing licenses:
        * if there are signed licenses already, use the existing status pubkey, and remove the trial activation flag from the status secret
        * if there are no signed licenses (but maybe unsigned ones), regenerate a private/public keypair

Reply:
You are right, there is another gap here.